src/codegen/compile.ml

```diff
@@ -3805,63 +3805,51 @@ module StableMem = struct
           E.else_trap_with env "StableMemory range out of bounds")
     | _ -> assert false
 
-  let read env name typ bytes load =
+  let add_guard env guarded get_offset bytes =
+    if guarded then
+     (if bytes = 1l then
+        get_offset ^^
+        guard env
+      else
+        get_offset ^^
+        compile_unboxed_const bytes ^^
+        guard_range env)
+    else G.nop
+
+  let read env guarded name typ bytes load =
     match E.mode env with
     | Flags.ICMode | Flags.RefMode ->
-      Func.share_code1 env ("__stablemem_read_" ^ name)
+      Func.share_code1 env ("__stablemem_" ^ (if guarded then "guarded_" else "") ^ "read_" ^ name)
         ("offset", I32Type) [typ]
         (fun env get_offset ->
           let words = Int32.div (Int32.add bytes 3l) 4l in
+          add_guard env guarded get_offset bytes ^^
           Stack.with_words env "temp_ptr" words (fun get_temp_ptr ->
             get_temp_ptr ^^ get_offset ^^  compile_unboxed_const bytes ^^
             IC.system_call env "ic0" "stable_read" ^^
             get_temp_ptr ^^ load))
     | _ -> assert false
 
-  let write env name typ bytes store =
+  let write env guarded name typ bytes store =
     match E.mode env with
     | Flags.ICMode | Flags.RefMode ->
-      Func.share_code2 env ("__stablemem_write_" ^ name)
+      Func.share_code2 env ("__stablemem_" ^ (if guarded then "guarded_" else "") ^ "write_" ^ name)
```

Comment:
I think using `sprintf` would be clearer here.

Reply:
Done.